### PR TITLE
Update spring

### PR DESCRIPTION
**Pull request by ericenns**
## Description of changes
Upgrade Spring to latest 4.X release, upgrade Spring Security to latest 4.X release, and upgrade spring-security-oauth2 to latest 2.X release.

## Related issue
Link to the GitHub issue this pull request addresses using the `#issuenum` format.  If it completes an issue, use `Fixes #issuenum` to automatically close the issue.

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.
* [ ] Tests added (or description of how to test) for any new features.
* [ ] User documentation updated for UI or technical changes.

### CODE REVIEW

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaOauthSecurityConfig.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+      endpoints.redirectResolver(new OauthRedirectResolver());
+    }
+
+    @Override
+    public void configure(final AuthorizationServerSecurityConfigurer oauthServer) throws Exception {
+      oauthServer
+        .tokenKeyAccess("permitAll()")
+        .checkTokenAccess("isAuthenticated()")
+        .allowFormAuthenticationForClients();
+      /*oauthServer.passwordEncoder(new PasswordEncoder() {
```
Comment (ericenns) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaOauthSecurityConfig.java]: I had copied this from your attempts at spring upgrade, I'm not sure if the password encoder is needed as we don't have a separate login endpoint for the api now.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaWebSecurityConfig.java [hunk 2]
```diff
@@ -197,26 +51,29 @@ public LoginSuccessHandler getLoginSuccessHandler() {
...
+			web.httpFirewall(new DefaultHttpFirewall());
 		}
 
 		@Override
 		protected void configure(HttpSecurity http) throws Exception {
 			authFailureHandler.setDefaultFailureUrl("/login?error=true");
 			// @formatter:off
 			http.requestMatcher(request -> {
-				return !request.getRequestURI().matches("^.*/api.*$");
+				return !request.getRequestURI().matches("^.*/api(?!/oauth/authorize|/oauth/authorization/token)(/.*)?$");
```
Comment (ericenns) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaWebSecurityConfig.java]: I had to modify this regex to allow `/api/oauth/authorize` and `/api/oauth/authorization/token` to be handled by this WebSecurity config as we need the specific web security context that the login form gives us.

File: src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaOauthSecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    @Override
+    public void configure(final ResourceServerSecurityConfigurer resources) {
+      resources.resourceId("NmlIrida").tokenServices(tokenServices);
+      forceExceptionTranslator(resources, exceptionTranslator);
+    }
+
+    @Override
+    public void configure(final HttpSecurity httpSecurity) throws Exception {
+      httpSecurity.antMatcher("/api/**").authorizeRequests()
+        //.antMatchers("/api/oauth/authorize").fullyAuthenticated()
```
Comment (ericenns) [inline src/main/java/ca/corefacility/bioinformatics/irida/config/security/IridaOauthSecurityConfig.java]: These antMatchers are commented out as they are being handled by the other WebSecurity config, see another comment below.

**Comment by ericenns** [comment]
@apetkau showed me how to run the sync process, I can confirm that refresh_tokens indeed work.

File: src/main/java/ca/corefacility/bioinformatics/irida/security/OauthRedirectResolver.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.springframework.security.oauth2.common.exceptions.InvalidRequestException;
+import org.springframework.security.oauth2.common.exceptions.OAuth2Exception;
+import org.springframework.security.oauth2.provider.ClientDetails;
+import org.springframework.security.oauth2.provider.endpoint.DefaultRedirectResolver;
+
+/**
+ * Redirect resolver for Oauth2 clients that allows any url for redirection
+ */
+public class OauthRedirectResolver extends DefaultRedirectResolver {
```
Comment (tom114) [inline src/main/java/ca/corefacility/bioinformatics/irida/security/OauthRedirectResolver.java]: We should mark this as a temporary fix while we're doing development.  We should have the redirect URIs registered on client creation.  Can you maybe `@Depreciated` this so we don't forget?

Comment (ericenns) [reply src/main/java/ca/corefacility/bioinformatics/irida/security/OauthRedirectResolver.java]: Added in f82f501296f49ac938eae69f6de8dda70d2daadc

**Pull request merged**